### PR TITLE
[Feature] Fp16 training for resnet50

### PR DESCRIPTION
**Pull request by typhoonzero**

### CODE REVIEW

**Review by chengduoZH — COMMENTED** [review]
You should add unit test for those fp16 kernel.

File: paddle/fluid/framework/details/scale_loss_grad_op_handle.cc [hunk 1]
```diff
@@ -22,39 +22,66 @@ namespace details {
...
+  OpHandleBase *op_handle_;
+  proto::VarType::Type out_dtype_;
+  platform::DeviceContext *ctx_;
+
+  ScaleLossGradFunctor(float coeff, Tensor *out, platform::Place place,
+                       OpHandleBase *op_handle, proto::VarType::Type dtype,
+                       platform::DeviceContext *ctx)
+      : coeff_(coeff), out_(out), place_(place), out_dtype_(dtype), ctx_(ctx) {}
+
+  template <typename OutT>
```
Comment (Yancey0623) [inline paddle/fluid/framework/details/scale_loss_grad_op_handle.cc:50]: Seems `OutT` => `T` is more common in this repo.

File: paddle/fluid/operators/optimizers/momentum_op.h [hunk 2]
```diff
@@ -109,13 +109,14 @@ class CPUDenseMomentumFunctor {
     auto p = framework::EigenVector<T>::Flatten(*param);
     auto v = framework::EigenVector<T>::Flatten(*velocity);
     auto g = framework::EigenVector<T>::Flatten(*grad);
-    auto* lr = learning_rate->data<T>();
+    const float* lr_f = learning_rate->data<float>();
```
Comment (chengduoZH) [inline paddle/fluid/operators/optimizers/momentum_op.h]: Why the `learning_rate` must be `float`?

Comment (Yancey0623) [inline paddle/fluid/operators/optimizers/momentum_op.h]: Add a comment to explain the reason?

Comment (typhoonzero) [reply paddle/fluid/operators/optimizers/momentum_op.h]: Will fix this. Currently learning rate created by lr scheduler is all float32

File: python/paddle/fluid/initializer.py [hunk 3]
```diff
@@ -207,16 +208,39 @@ def __call__(self, var, block):
...
+                type=VarDesc.VarType.LOD_TENSOR,
+                persistable=False)
+        else:
+            out_dtype = var.dtype
+            out_var = var
+
         op = block._prepend_op(
             type="uniform_random",
-            outputs={"Out": var},
+            outputs={"Out": out_var},
```
Comment (chengduoZH) [inline python/paddle/fluid/initializer.py:227]: We should add fp16 kernle for `uniform_random`.

File: paddle/fluid/operators/fill_zeros_like_op.cu.cc [hunk 4]
```diff
@@ -22,4 +23,6 @@ REGISTER_OP_CUDA_KERNEL(
     ops::FillZerosLikeKernel<paddle::platform::CUDADeviceContext, int64_t>,
     ops::FillZerosLikeKernel<paddle::platform::CUDADeviceContext, float>,
     ops::FillZerosLikeKernel<paddle::platform::CUDADeviceContext, double>,
+    ops::FillZerosLikeKernel<paddle::platform::CUDADeviceContext,
```
Comment (chengduoZH) [inline paddle/fluid/operators/fill_zeros_like_op.cu.cc:26]: You had better add `InferVarType` for `fill_zeros_like `, otherwise the dtype of the output var maybe not fp16.

File: paddle/fluid/operators/metrics/accuracy_op.cu [hunk 5]
```diff
@@ -94,6 +95,7 @@ class AccuracyOpCUDAKernel : public framework::OpKernel<T> {
 
 // FIXME(typhoonzero): types of T is for inference data.
 // label data is always int64
-REGISTER_OP_CUDA_KERNEL(accuracy,
-                        paddle::operators::AccuracyOpCUDAKernel<float>,
-                        paddle::operators::AccuracyOpCUDAKernel<double>);
+REGISTER_OP_CUDA_KERNEL(
+    accuracy, paddle::operators::AccuracyOpCUDAKernel<float>,
+    paddle::operators::AccuracyOpCUDAKernel<double>,
+    paddle::operators::AccuracyOpCUDAKernel<paddle::platform::float16>);
```
Comment (chengduoZH) [inline paddle/fluid/operators/metrics/accuracy_op.cu:101]: Had better add `InferVarType` for `accuracy` to infer the output var's dtype.

**Review by chengduoZH — APPROVED** [review]
I have add some comments for this PR, but it can be merged first.

File: paddle/fluid/operators/metrics/accuracy_op.cu (hunk 5 above)
Comment (typhoonzero) [reply paddle/fluid/operators/metrics/accuracy_op.cu:101]: Thanks, moving changes to https://github.com/PaddlePaddle/Paddle/pull/14899 WIP.
